File: cpp/test/encoder_test.cpp
#define CATCH_CONFIG_MAIN
#include <catch2/catch.hpp>

#include <fstream>
#include <string>

#include <nlohmann/json.hpp>

#include "spanvalue/encoder.hpp"

namespace {

using Json = nlohmann::json;

Json load_conformance(const std::string& path) {
  std::ifstream in(path);
  REQUIRE(in.is_open());
  Json data;
  in >> data;
  return data;
}

}  // namespace

#ifndef SPANVALUE_CONFORMANCE_PATH
#define SPANVALUE_CONFORMANCE_PATH "../testdata/conformance.json"
#endif

TEST_CASE("encoder round-trip value cases", "[encoder]") {
  const Json conformance = load_conformance(SPANVALUE_CONFORMANCE_PATH);
  const spanvalue::FormatConfig config = spanvalue::simple_format_config();

  for (const Json& case_obj : conformance.at("value_cases")) {
    const Json& typ = case_obj.at("type");
    const Json& wire = case_obj.at("value");
    const Json native = spanvalue::wire_to_native(typ, wire);
    const Json encoded = spanvalue::encode_value(typ, native);
    INFO("encoder case " << case_obj.at("name").get<std::string>());
    REQUIRE(spanvalue::wire_equal(encoded, wire));

    const std::string formatted = spanvalue::format_value(typ, encoded, config);
    const std::string want = case_obj.at("expected").at("simple");
    REQUIRE(formatted == want);
  }
}

TEST_CASE("format_result_row smoke", "[encoder]") {
  const Json bool_type = {{"code", "BOOL"}};
  const Json int_type = {{"code", "INT64"}};
  const Json str_type = {{"code", "STRING"}};
  const std::vector<Json> types = {bool_type, int_type, str_type};
  const std::vector<Json> natives = {true, nullptr, "ok"};
  const spanvalue::FormatConfig config = spanvalue::simple_format_config();

  const std::vector<std::string> got = spanvalue::format_result_row(types, natives, config);
  REQUIRE(got.size() == 3);
  REQUIRE(got[0] == "true");
  REQUIRE(got[1] == "<null>");
  REQUIRE(got[2] == "ok");
}

TEST_CASE("format_result_row struct and null", "[encoder]") {
  const Json struct_type = {
      {"code", "STRUCT"},
      {"structType",
       {{"fields",
         Json::array({{{"name", "n"}, {"type", {{"code", "INT64"}}}},
                      {{"name", "s"}, {"type", {{"code", "STRING"}}}}})}}}};
  const Json bool_type = {{"code", "BOOL"}};
  const std::vector<Json> types = {bool_type, struct_type};
  const std::vector<Json> natives = {nullptr, Json::array({42, "x"})};
  const spanvalue::FormatConfig config = spanvalue::simple_format_config();

  const std::vector<std::string> got = spanvalue::format_result_row(types, natives, config);
  REQUIRE(got.size() == 2);
  REQUIRE(got[0] == "<null>");
  REQUIRE(got[1] == "(42 AS n, x AS s)");
}

File: php/tests/EncoderTest.php
<?php

declare(strict_types=1);

namespace Apstndb\SpanValue\Tests;

use Apstndb\SpanValue\Encoder;
use Apstndb\SpanValue\ValueFormat;
use PHPUnit\Framework\TestCase;

final class EncoderTest extends TestCase
{
    private static ?array $conformance = null;

    public static function setUpBeforeClass(): void
    {
        self::$conformance = json_decode(
            (string) file_get_contents(dirname(__DIR__, 2) . '/testdata/conformance.json'),
            true,
            512,
            JSON_THROW_ON_ERROR
        );
    }

    private static function conformance(): array
    {
        return self::$conformance ?? throw new \RuntimeException('conformance data not loaded');
    }

    public function testEncodeRoundTripFromConformance(): void
    {
        foreach (self::conformance()['value_cases'] as $case) {
            $type = $case['type'];
            $wire = $case['value'];
            $native = Encoder::wireToNative($type, $wire);
            $encoded = Encoder::encodeValue($type, $native);
            self::assertTrue(
                Encoder::wireEqual($type, $encoded, $wire),
                "encode round-trip {$case['name']}: encoded "
                . var_export($encoded, true)
                . ' want ' . var_export($wire, true)
            );
        }
    }

    /** @return list<array{0: string}> */
    public static function valuePresetProvider(): array
    {
        return array_map(static fn (string $p) => [$p], ['simple', 'literal', 'spanner_cli']);
    }

    /** @dataProvider valuePresetProvider */
    public function testEncodeThenFormatMatchesConformance(string $preset): void
    {
        $config = match ($preset) {
            'simple' => ValueFormat::simpleFormatConfig(),
            'literal' => ValueFormat::literalFormatConfig(),
            'spanner_cli' => ValueFormat::spannerCliFormatConfig(),
            default => throw new \InvalidArgumentException("unknown preset {$preset}"),
        };

        foreach (self::conformance()['value_cases'] as $case) {
            $type = $case['type'];
            $native = Encoder::wireToNative($type, $case['value']);
            $encoded = Encoder::encodeValue($type, $native);
            $got = ValueFormat::formatValue($type, $encoded, $config);
            $want = $case['expected'][$preset];
            self::assertSame(
                $want,
                $got,
                "encode+format {$case['name']} preset {$preset}: got " . var_export($got, true)
                . ' want ' . var_export($want, true)
            );
        }
    }

    public function testFormatResultRow(): void
    {
        $types = [
            ['code' => 'INT64'],
            [
                'code' => 'STRUCT',
                'structType' => [
                    'fields' => [
                        ['name' => 'n', 'type' => ['code' => 'INT64']],
                        ['name' => 's', 'type' => ['code' => 'STRING']],
                    ],
                ],
            ],
            ['code' => 'STRING'],
        ];
        $nativeValues = [42, ['n' => 1, 's' => 'hello'], null];
        $config = ValueFormat::simpleFormatConfig();

        $got = Encoder::formatResultRow($types, $nativeValues, $config);
        self::assertSame(['42', '(1 AS n, hello AS s)', '<null>'], $got);
    }

    public function testStructAssociativeEncoding(): void
    {
        $type = [
            'code' => 'STRUCT',
            'structType' => [
                'fields' => [
                    ['name' => 'a', 'type' => ['code' => 'INT64']],
                    ['name' => 'b', 'type' => ['code' => 'BOOL']],
                ],
            ],
        ];
        $encoded = Encoder::encodeValue($type, ['a' => '7', 'b' => true]);
        self::assertSame(['7', true], $encoded);
    }

    public function testBytesFromRawBinary(): void
    {
        $type = ['code' => 'BYTES'];
        $encoded = Encoder::encodeValue($type, "\x00ok");
        self::assertSame('AG9r', $encoded);
    }

    public function testJsonFromArray(): void
    {
        $type = ['code' => 'JSON'];
        $encoded = Encoder::encodeValue($type, ['x' => 1]);
        self::assertSame('{"x":1}', $encoded);
    }
}
